back/src/main/java/com/ubikloadpack/jmeter/ulp/observability/util/ErrorsMap.java

```diff
@@ -0,0 +1,102 @@
+	
+	public ErrorsMap() {
+		this.errorsPerType = new ConcurrentHashMap<>();
+	}
+
+	/**
+	 * Add an error type and increment it's occurrence.
+	 * @param errorType the error type to add.
+	 * @return The information of the added error type.
+	 */
+	public ErrorTypeInfo addErrorTypeAndCount(String errorType) {
+		ErrorTypeInfo errorTypeInfo = this.errorsPerType.computeIfAbsent(errorType, s -> new ErrorTypeInfo(errorType));
+	    errorTypeInfo.increment();
+	    return errorTypeInfo;
+	}
+	
+	/**
+	 * Compute the frequency of that error type.
+	 * @param errorType the type of error for which the frequency will be calculated
+	 * @param totalErrors the total errors. Should get it's value from the Micrometer registry.
+	 * @return Percentage of occurrences of a specific error type out of the total number of errors.
+	 * Returns null if no errorType key was found.
+	 */
+	public Double computeErrorTypeFrequency(String errorType, Long totalErrors) {
+		if (this.errorsPerType.contains(errorType)) {
+			return errorsPerType.get(errorType).computeErrorTypeFrequency(totalErrors);
+		}
+		return null;
+	}
+	
+	/**
+	 * Compute error rate of that error type.
+	 * @param errorType the type of error for which the error rate will be calculated.
+	 * @param totalRequests the total requests. Should get it's value from the Micrometer registry.
+	 * @return percentage of queries that raised an error of this type. Returns null if no errorType 
+	 * key was found.
+	 */
+	public Double computeErrorRateForType(String errorType, Long totalRequests) {
+		if (this.errorsPerType.contains(errorType)) {
+			return errorsPerType.get(errorType).computeErrorRate(totalRequests);
+		}
+		return null;
+	}
+	
+	/**
+	 * Build an ErrorsMap that contains the X top Errors.
+	 * @param maxErrors The maximum type errors to keep.
+	 * @return a new ErrorMap with only the top X Errors.
+	 */
+	public ErrorsMap collectTopXErrors(int maxErrors) {
+	    Map<String, ErrorTypeInfo> topErrors = this.errorsPerType.entrySet()
+	            .stream()
+	            .sorted(Map.Entry.<String, ErrorTypeInfo>comparingByValue().reversed())
+	            .limit(maxErrors)
+	            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+	    ErrorsMap result = new ErrorsMap();
+	    result.errorsPerType = new ConcurrentHashMap<>(topErrors);
+	    return result;
+	}
+	
+	/**
+	 * Get the openMetric format of the error types.
+	 * @param sampleName the name of the sample (should be total_label)
+	 * @param totalThreads the number of the total threads
+	 * @param totalErrors
+	 * @param timeStamp 
+	 * @return
+	 */
+	public String toOpenMetric(String sampleName, Long totalThreads, Long totalErrors, Long timeStamp) {
+		StringBuilder str = new StringBuilder();
+		
+		this.errorsPerType.forEach((errorType, errorInfo) -> {
+			Double errorFrequency = errorInfo.computeErrorTypeFrequency(totalErrors);
+			Double errorRate = errorInfo.computeErrorRate(totalThreads);

```

Comment:
Is it totalThreads to requestsTotal ?

Reply:
ah! it's not. This is the total number of threads. I'm doing the wrong calculation. I was thinking of calculating the percentage of threads that raise this type of error. But the objective is to calculate the percentage of the type of error on all requests.